Symex: resolve pointer comparisons using the value-set

owen-mc-diffblue opened the pull request:
This allows us to evaluate boolean expressions (including GOTO conditions) that test pointer equality or inequality (particularly "ptr == null" or "!= null") where the value-set contains sufficient information. We can already handle simple cases where `ptr` has a single known alias via constant propagation and aliasing, but not cases where `ptr` is non-constant, for example because it is drawn from an array.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add tests for symex evaluating simple pointer comparisons

At time of writing, develop fails on the following lines of the test:
test::1::unconditionally_reachable_1[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_1[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_6[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_6[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_7[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_7[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_8[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_8[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_9[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_9[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_10[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_10[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_11[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_11[^\s]+\)$ [FAILED]
test::1::unconditionally_reachable_12[^\s]+ = \(goto_symex::\\guard[^\s]+ \? 7 : test::1::unconditionally_reachable_12[^\s]+\)$ [FAILED]
test::1::unreachable_1[^\s]+ = 7$ [FAILED]
test::1::unreachable_6[^\s]+ = 7$ [FAILED]
test::1::unreachable_7[^\s]+ = 7$ [FAILED]
test::1::unreachable_8[^\s]+ = 7$ [FAILED]
test::1::unreachable_9[^\s]+ = 7$ [FAILED]
test::1::unreachable_10[^\s]+ = 7$ [FAILED]
test::1::unreachable_11[^\s]+ = 7$ [FAILED]
test::1::unreachable_12[^\s]+ = 7$ [FAILED]

owen-mc-diffblue commented on src/goto-symex/goto_symex.h:
```diff
@@ -287,6 +287,10 @@ class goto_symext
     statet &,
     bool keep_array);
 
+  exprt try_evaluate_pointer_comparisons(
```
Any suggestions for a better name?

smowton replied:
I like this one!

owen-mc-diffblue commented on src/goto-symex/goto_symex.h:
```diff
@@ -333,6 +337,11 @@ class goto_symext
     const exprt &new_guard,
     const namespacet &ns);
 
+  optionalt<exprt> build_reference_to_value_set_element(
```
This function might fit better somewhere else - perhaps in `value_set_dereferencet`, since it's mainly a wrapper around `value_set_dereferencet::build_reference_to`?

smowton replied:
Proooobably. No strong opinion.

owen-mc-diffblue commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
+      op = try_evaluate_pointer_comparisons(std::move(op), value_set);
+    }
+  }
+  else if(condition.id() == ID_not)
```
anything else I should be recursing over, other than AND, OR and NOT?

smowton replied:
No I think that's enough

smowton replied:
Oh wait, how about the arguments of a boolean-typed `if_exprt`

smowton replied:
(e.g. `x == y ? z == w : a == b`)

smowton replied:
Or indeed the first arg of an `if_exprt` in any context. Total recursion looking for `==` and `!=` nodes might be the simplest thing.

smowton replied:
I 100% think these comments through before beginning to write FYI

owen-mc-diffblue commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
+    {
+      symbol_exprt *symbol_expr =
+        expr_try_dynamic_cast<symbol_exprt>(condition.op0());
+      exprt &constant = condition.op1();
```
Currently it assumes the symbol is on the lhs and the constant is on the rhs. I will fix that up.

owen-mc-diffblue commented on src/goto-symex/symex_main.cpp:
```diff
@@ -583,6 +583,38 @@ find_unique_pointer_typed_symbol(const exprt &expr)
   return return_value;
 }
 
+optionalt<exprt> goto_symext::build_reference_to_value_set_element(
```
This code was in `goto_symext::try_filter_value_sets` and I've pulled it out into a function to reuse it. Perhaps I'll make that a separate commit.

smowton commented on regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c:
```diff
@@ -0,0 +1,76 @@
+    possibly_reachable_12 = 7;
+
+  int unconditionally_reachable_13;
+  if((ptr_to_a == &a && c == 3) || c == 0)
```
Bear in mind these constant comparisons with `c` may well be resolved very early as `c` is only assigned once and never address-taken.

owen-mc-diffblue replied:
I've changed them to other pointer comparisons. And I now take the address of all the integers at the beginning, if that's a thing that symex cares about.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
+        expr_try_dynamic_cast<symbol_exprt>(condition.op0());
+      exprt &constant = condition.op1();
+
+      if(symbol_expr && goto_symex_is_constantt()(constant))
```
Handle the reverse case too (constant OP symbol)

smowton replied:
(and test this)

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
+      if(symbol_expr && goto_symex_is_constantt()(constant))
+      {
+        value_setst::valuest value_set_elements;
+        value_set.get_value_set(
```
Assert that `symbol_expr` is already L1 at least

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
+          to_ssa_expr(*symbol_expr).get_l1_object(), value_set_elements, ns);
+
+        INVARIANT(
+          constant.id() == ID_address_of || constant.id() == ID_NULL,
```
Surprising! Null is usually represented as an `ID_constant` with value = NULL -- suspect you need to add tests for this

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
+
+        if(!unknown_or_invalid_found)
+        {
+          if(!constant_found)
```
Give some brief comments setting out the logic here, as `!constant_found` and `value_set_elements.size() == 1` don't immediately chime with "x in { a, b, c }" and "x in { x }" respectively.

smowton reviewed (commented):
Looks good except I believe the null case is broken, and at least needs tests.

Commit:
Add pointer field to value_set_dereferencet::valuet

value_set_dereferencet::valuet is only used as the return value of
value_set_dereferencet::build_reference_to. valuet::value is an
expression for the object pointed to, which is what's needed for
dereferencing. In other contexts, e.g. filtering value-sets, what
is needed is an expression for the pointer. This logic already
existed in try_filter_value_set, but putting it directly in
build_reference_to makes it easier to use it in other places and
makes it clear that it didn't cope with typecasts properly.

Commit:
Evaluate some pointer comparisons as true or false

Use the value-set to evaluate pointer comparisons as true or false
where possible

owen-mc-diffblue commented:
I've addressed all of @smowton's review comments on the draft PR and got CI working (touch wood).

smowton commented on regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c:
```diff
@@ -0,0 +1,188 @@
+
+  // Equality "=="
+
+  // A non-null pointer and the correct value
```
correct -> matching and incorrect -> non-matching everywhere

smowton commented on regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c:
```diff
@@ -0,0 +1,188 @@
+
+  // Disequality "!="
+
+  // A non-null pointer and the correct value
```
the -> a

smowton commented on regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.desc:
```diff
@@ -0,0 +1,51 @@
+test::1::unreachable_12[^\s]+ = 7$
+^warning: ignoring
+--
+Pointer comparisons will be resolved in symex by a mixture of constant
```
Suggest adding conditions that check you check for everything you should, by making sure `unreachable_1[3-9])` and `unreachable[2-9][0-9]` for example don't occur anywhere (i.e., if someone added a test but not a cross-check this regex would complain)

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,182 @@ void goto_symext::apply_goto_condition(
+
+  INVARIANT(
+    other_operand.id() == ID_address_of ||
+      (typecast_expr && typecast_expr->op().id() == ID_address_of) ||
```
If you're not sure this is simplified you might get redundant stacked typecast ops

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,182 @@ void goto_symext::apply_goto_condition(
+    expr_try_dynamic_cast<symbol_exprt>(symbol_expr);
+
+  if(!symbol_expr_lhs)
+  {
```
No need for braces with one-liner `if`s

owen-mc-diffblue replied:
I prefer to use them in all cases

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,182 @@ void goto_symext::apply_goto_condition(
+/// \param value_set: The value-set for looking up what the symbol can point to
+/// \param language_mode: The language mode
+/// \param ns: A namespace
+static bool try_evaluate_pointer_comparisons_base_case_with_check(
```
Instead of calling this twice, suggest something like https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_state.cpp#L67

smowton replied:
I don't think there's much gained by the two functions (with_check vs. without) since they're both only called in one place

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,182 @@ void goto_symext::apply_goto_condition(
+/// \param value_set: The value-set for looking up what the symbol can point to
+/// \param language_mode: The language mode
+/// \param ns: A namespace
+static void try_evaluate_pointer_comparisons_base_case(
```
How about `try_evaluate_pointer_comparison`, singular?

owen-mc-diffblue commented:
Note also that the last two commits break sharing everywhere, so we'll have to rethink how to do that.

smowton commented:
Suggested change: https://github.com/smowton/cbmc/commit/1daa2c60056213d2eddce6f9e3a2945a139d81b2

That removes `renamed_vectort` and instead uses `depth_begin/depth_end`, which avoid sharing breaks when no changes are made, to selectively alter the renamed expression, only permitting changes where the renaming level is consistent. @romainbrenguier are you happy with this?

@owen-jones-diffblue note that also applies a couple of my suggestions above since it made it easier to code :)

romainbrenguier commented:
@smowton 
> Suggested change: smowton@1daa2c6
> That removes renamed_vectort and instead uses depth_begin/depth_end, which avoid sharing breaks when no changes are made, to selectively alter the renamed expression, only permitting changes where the renaming level is consistent. @romainbrenguier are you happy with this?

Yes that looks better. What I'm not sure about is the fact that the mutator function takes an `exprt` as argument rather that a `renamedt`. That way you avoid calling constructors on all intermediary expressions, but I don't think it's nice for the interface. Maybe we could make `renamedt` privately inherit from exprt so that we can use `static_cast` instead of a constructor, so that conversion would be cheap and the interface nicer?
Here is the change I propose (if you want to use it): https://github.com/diffblue/cbmc/pull/4454

owen-mc-diffblue commented:
Thanks @smowton, that's much better. I've cherry-picked it and squashed it in, along with fixes for all the things you mentioned. I did add an extra commit which makes `selectively_mutate` a free function (which is a friend of `renamedt`), since it doesn't make sense to be using depth iterators for anything when `underlyingt` is derived from `exprt`. I'm not sure how this interacts with @romainbrenguier's suggestion.

There are three failing tests. The first is `exception_cleanup`, which is failing because symex can now tell that `@inflight_exception` is not null, so the test needs to be changed. I'm not totally sure what it should now be. The failing lines are:
```
1 remaining after simplification$
```
in test.desc and
```
file Test\.java line 6 function java::Test.main:\(I\)V
no uncaught exception
```
in vccs.desc.

The other two are nondet_elements_longer_lists and nondet_elements_longer_lists_global, which are failing in the same way. I'm looking into that at the moment.

owen-mc-diffblue commented:
For nondet_elements_longer_lists and nondet_elements_longer_lists_global, the test is failing to find these two lines: ```
\[test_function.unwind.\d+\] line \d+ unwinding assertion loop 0: SUCCESS
\[test_function.unwind.\d+\] line \d+ unwinding assertion loop 1: SUCCESS
```
Is there a good reason that we would not generate these unwinding assertions? If symex can determine that they are true then does it just not output them in the formula?

smowton commented:
@owen-jones-diffblue unsure about the unwinding assertions (yet) but the exception one is entirely desirable -- that 1 VCC remaining was a shortcoming which we have now fixed (symex can tell no exception escaped from the whole program). Just amend the test to check *all* conditions are symex'd away.

smowton commented:
@romainbrenguier I'm not sure I follow, what would your ideal interface look like?

tautschnig commented:
@owen-jones-diffblue Presumably your changes just result in more constant propagation, and thus those unwinding assertions are resolved by the simplifier already. #2574 does the same, i.e., removes those lines from the test spec.

romainbrenguier commented:
@smowton 
> @romainbrenguier I'm not sure I follow, what would your ideal interface look like?

I would think having `mutator_functiont` take `renamedt` arguments would be more useful (though in your case it doesn't really matter). For instance imagine we want to selectively apply `simplify` to operands of a `renamedt`, you could do it with a `renamedt -> optinalt<renamedt>` function. I'm not insisting on it, since it is not necessary here, but could be nice in the future.

smowton commented:
Ah sure, ok. On the one hand the tag is useful if you want to call a function that demands a `renamedt`, but I certainly like not having to make a dedicated object for every subexpression, so your privately-inherit trick looks like a good solution.

Commit:
Update existing tests so they pass

owen-mc-diffblue commented:
I've marked this as `do not merge` until we have some benchmarking data for it

owen-mc-diffblue commented:
I've cherry-picked ##4454 and added a commit to use it. Before merging this should be cleaned up.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eb9b8cecade7195d17aa7de83276962168a60daa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106349820

tautschnig commented on src/util/ssa_expr.h:
```diff
@@ -171,4 +171,14 @@ inline bool is_ssa_expr(const exprt &expr)
+inline void validate_expr(const ssa_exprt &value)
+{
+  validate_operands(value, 0, "SSA expressions must not have operands");
+}
```
If adding those, could we please unify the various checking-functions in the same way this has been implemented in `std_expr.h` for all the classes? That is, use `validate_expr` in the above.

tautschnig commented on src/pointer-analysis/value_set_dereference.h:
```diff
@@ -74,6 +74,13 @@ class value_set_dereferencet final
     irep_idt language_mode,
     const namespacet &ns);
 
+  static optionalt<exprt> build_reference_to_value_set_element(
```
Please document.

owen-mc-diffblue replied:
What I want to call it is `build_reference_to`, but that's already taken by the function it's wrapping (which doesn't actually build a reference).

The reason I have a wrapper function is that `build_reference_to` makes it easy to get objects, but not pointers, e.g. it gives you `global`, not `&global`, and for null pointers it uses the special value `nil_exprt`, and doesn't give me the `null_pointer_exprt` that I want. That makes sense because I think it's only used for `dereference_rec`, which only cares about dereferenced pointers.  Now that I've read through that function a few more times, I think a better solution would be to add a field to `value_set_dereferencet::valuet` to put `&global` and `null_pointer_exprt` into. What would you say to that?

owen-mc-diffblue replied:
I have done this now

tautschnig commented on regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.desc:
```diff
@@ -0,0 +1,60 @@
+propagation and value-set filtering in try_evaluate_pointer_comparisons.
+
+The expected failure for path-symex is because the lines we check for
+possibly_reachable_* would only appear when there is convergence behaviour.
```
GitHub suggests adding a final newline.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,167 @@ void goto_symext::apply_goto_condition(
+    other_operand_with_typecasts_removed = &typecast_expr->op();
+    typecast_expr = expr_try_dynamic_cast<typecast_exprt>(
+      *other_operand_with_typecasts_removed);
+  }
```
Use `skip_typecast`?

owen-mc-diffblue replied:
I tried to search for such a utility but didn't hit the right combination of words. Next time I'll try looking in `expr_utils.h`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 926531eae1c1eeb6c03c9deab5570c422639484e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106591286

owen-mc-diffblue commented:
@tautschnig I've addressed all but one comment. Let me know what you think about my suggested approach to addressing the last comment.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 44d45d645e96fc5787b21c97e20ca4e5e4f400a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106623135

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c0e7ab8c6151a9943d901e145acffc7a581c10a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106938073

smowton commented:
@tautschnig initial findings from studying this: the test `jbmc/regression/jbmc/exception-cleanup` gives a good example of where this can help, as even with #3619 and #4430 we get an expression where propagation works, but the simplifier isn't clever enough to spot that a unique value can be determined -- in this case we get `guard ? null : (guard ? &x : null)`, which is clearly always null but requires the simplifier to propagate that `guard => false` into the RHS of the conditional, which it doesn't attempt at present.

Here's another example where propagation alone doesn't work:
```c
#include <assert.h>

int main(int argc, char **argv) {
  int x, y;
  int *p = argc == 1 ? 0 : argc == 2 ? &x : &y;
  if(p != 0) {
    if(p == 0) {
      assert(0);
    }
  }
}
```

The redundant null-tests on `p` can't be resolved because by simplification because the knowledge that the null case has already been filtered out is encoded in the value-set but not in the expression itself. This could be done, for example by transforming `p == (argc == 1 ? 0 : argc == 2 ? &x : &y) && p != 0` into `argc == 2 ? &x : &y`, but we currently don't attempt this, and if we did I expect we would either (a) need to switch on local-definitions in the simplifier, which could get quite expensive, or (b) the augmented simplifier would probably only catch particular cases rather than generally propagating conditions into expressions.

In brief I continue to think that at least for the time being the value-set is a more robust way of tracking aliases than trying to propagate all if-expressions encoding aliases and improve the simplifier sufficiently to filter those nested conditionals and solve tests against them. I think this is particularly a problem as in the limiting case this results in spiralling expression sizes and so the cost of `simplify` and `merge_ireps` rises, as I'm currently trying to fight *against* in #4576.

tautschnig commented on src/goto-symex/renaming_level.h:
```diff
@@ -98,12 +103,51 @@ class renamedt : private underlyingt
+  friend void selectively_mutate(
+    renamedt<exprt, selectively_mutate_level> &renamed,
+    typename renamedt<exprt, selectively_mutate_level>::mutator_functiont
+      get_mutated_expr);
```
I'm a bit surprised by the amount of friendship in here. Couldn't some of these be made static members instead?

smowton replied:
The gymnastics here are a little beyond me, so I think I'll leave this as-is and welcome any cleanup PRs by people who're better at this game than me :)

tautschnig reviewed (approved):
Looks ok me, except the remaining clang-format problem and, more importantly, the absence of performance data - could this please be provided? I certainly see that this gives symex additional power in resolving some comparisons, but at what cost does this come?

smowton commented:
cbmc CORE unit tests times: unchanged

smowton commented:
Large Webgoat example: 55 seconds to reach a particular checkpoint (down 5 seconds)

smowton commented:
Romain's Sakai benchmark: One benchmark reduced around 15 seconds (from 145 seconds), the remainder tested showed negligible difference.

smowton commented:
Therefore I'm pretty confident this has no noticeable cost in context of the other fairly-expensive ops we perform in symex.

peterschrammel commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,161 @@ void goto_symext::apply_goto_condition(
+  const symbol_exprt &symbol_expr,
+  const exprt &other_operand,
+  const value_sett &value_set,
+  const irep_idt language_mode,
```
`const &`

peterschrammel commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,161 @@ void goto_symext::apply_goto_condition(
+/// \return If we were able to evaluate the condition as true or false then we
+///   return that, otherwise we return an empty optionalt
+static optionalt<renamedt<exprt, L2>> try_evaluate_pointer_comparison(
+  irep_idt operation,
```
`const &`

peterschrammel commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,161 @@ void goto_symext::apply_goto_condition(
+static optionalt<renamedt<exprt, L2>> try_evaluate_pointer_comparison(
+  const renamedt<exprt, L2> &renamed_expr,
+  const value_sett &value_set,
+  const irep_idt language_mode,
```
`const &`

peterschrammel commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -65,6 +67,161 @@ void goto_symext::apply_goto_condition(
+static renamedt<exprt, L2> try_evaluate_pointer_comparisons(
+  renamedt<exprt, L2> condition,
+  const value_sett &value_set,
+  const irep_idt language_mode,
```
`const &`

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ff6002bb92522460931c3ffc6c3d517e3ac8b0b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111618837

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 213a1b25c0eeb10405005619b45ce71eb82490a1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111750931

tautschnig commented:
I'm afraid this actually causes spurious counterexamples being reported. I'll try to build a small regression test.

tautschnig commented:
Actually it's pretty simple:
```
int main()
{
  void *ptr;
  if(ptr == 0)
  {
    return 0;
  }
  __CPROVER_assert(ptr != 0, "null");
}
```
I believe the problem is that `ptr` has never been assigned, and thus the value set has no information about it.

smowton commented:
Discussion on Slack summary: will revert and re-submit with a fix
